src/dilithium.cpp

```diff
@@ -0,0 +1,448 @@
+{
+    return std::make_unique<DilithiumSpec>(getType());
+}
+
+DilithiumAsymmetricKeypair DilithiumAsymmetricKeypair::generate(
+        const DilithiumAsymmetricKey::Spec &spec)
+{
+    return DilithiumAsymmetricKeypair(spec.generate());
+}
+
+DilithiumAsymmetricKey DilithiumSpec::generate() const
+{
+    std::vector<uint8_t> privateKey(getPrivKeySize(_keyType));
+    // The public key is expected. Nullptr would lead to errors
+    // So we need to create a vector although we throw away the value later.
+    std::vector<uint8_t> publicKey(getPubKeySize(_keyType));
+
+    switch (_keyType) {
+        case KeyTypes::DILITHIUM2:
+            if (pqcrystals_dilithium2_ref_keypair(publicKey.data(), privateKey.data())) {
+                throw MoCOCrWException("Cannot create dilithium keypair.");
+            }
+            break;
+        case KeyTypes::DILITHIUM3:
+            if (pqcrystals_dilithium3_ref_keypair(publicKey.data(), privateKey.data())) {
+                throw MoCOCrWException("Cannot create dilithium keypair.");
+            }
+            break;
+        case KeyTypes::DILITHIUM5:
+            if (pqcrystals_dilithium5_ref_keypair(publicKey.data(), privateKey.data())) {
+                throw MoCOCrWException("Cannot create dilithium keypair.");
+            }
+            break;
+        default:
+            throw MoCOCrWException("The used dilithium key type is not supported.");
+    }
+
+    return DilithiumAsymmetricKey(std::make_shared<DilithiumKeyImpl>(privateKey, _keyType));
+}
+
+template <class Key>
+class DilithiumSignatureImpl
+{
+public:
+    DilithiumSignatureImpl(const Key &key) : _key(key) {}
+
+protected:
+    Key _key;
+};
+
+class DilithiumSigningCtx::Impl : public DilithiumSignatureImpl<DilithiumAsymmetricPrivateKey>
+{
+public:
+    using DilithiumSignatureImpl<DilithiumAsymmetricPrivateKey>::DilithiumSignatureImpl;
+    std::vector<uint8_t> signMessage(const std::vector<uint8_t> &message)
+    {
+        std::vector<uint8_t> signature;
+        size_t signatureLength;
+        std::function<int(uint8_t *, size_t *, const uint8_t *, size_t, const uint8_t *)>
+                sign_function;
+        switch (_key.getType()) {
+            case KeyTypes::DILITHIUM2:
+                signature.resize(pqcrystals_dilithium2_BYTES);
+                sign_function = pqcrystals_dilithium2_ref_signature;
+                break;
+            case KeyTypes::DILITHIUM3:
+                signature.resize(pqcrystals_dilithium3_BYTES);
+                sign_function = pqcrystals_dilithium3_ref_signature;
+                break;
+            case KeyTypes::DILITHIUM5:
+                signature.resize(pqcrystals_dilithium5_BYTES);
+                sign_function = pqcrystals_dilithium5_ref_signature;
+                break;
+            default:
+                throw MoCOCrWException("Invalid key type for dilithium");
+        }
+
+        if (sign_function(
```

Comment:
I wanted to keep the switch/case part as small as possible and tried to avoid code-duplication. But I am also fine with moving code into the switch/case statement.